design/baremetal-operator/raid-disk-controller-naming.md

```diff
@@ -0,0 +1,213 @@
+and RAID controllers within Ironic. This leads to a complication for the
+operator where user experience is affected. There needs to be a uniform
+naming convention that abstracts this and increases convenience.
+
+### Goals
+
+The primary goal is to reach consensus on one naming convention to depict
+disk and RAID controller names in the baremetal-host ``hardware-raid``
+section.
+
+Afterwards, implementation of the same will be done within the
+``baremetal-operator`` by extending the ``baremetal-host`` specification.
+
+### Non-Goals
+
+This specification does not deal with ``software-raid`` and extension for it.
+
+## Proposal
+
+The crux of the proposal is the naming convention that is to be adopted
+for interfacing with the user. Currently, we have the following options
+
+### Controller Naming
+
+For controllers that are installed specifically for the operating system,
+for example the [BOSS Card](1) by DellEMC, they are almost always used with
+two disks/drives in ``RAID-1``, and can be refered to as
+``primary-controller``. Similar method can be adopted to address such
+controllers by other vendors.
+
+Other controllers that are installed in the server can be used in any order,
+and thus can be named ``secondary-controller-x`` where x can be an integer
+greater than or equal to 1, e-g ``secondary-controller-1``,
+``secondary-controller-2`` and so on.
+
+### Physical Disk/Drive Naming
+
+With physical disks, the primary issue is disk enumeration. We can address
+it this way:
+
+- Start from the front bay, and count drives in the order that they are
+  listed by vendors. For instance, vendors would list them as XXXXX-1-XXX,
+  XXXXX-2-XXX and so on. We can use these 1 and 2 to specify ordering and
+  provide a simpler naming convention as ``Disk-1``, ``Disk-2`` and so on.
+- The drives inserted in the backplane can be counted _afterwards_ using
+  the same convention.
+
+This will provide the user with a simpler way of listing controllers
+and disks and it should suffice for a significant majority of use-cases.
+
+### User Stories
+
+The following user stories appertain to the proposal in question
+
+#### Story 1
+
+As an operator, I should be able to focus on defining my RAID configurations
+without having to deal with intricacies of individual vendor's naming
+conventions.
+
+#### Story 2
+
+As an operator, I'd like to be able to specify the physical disks I want
+to use for my RAID configuration. Or, in case of multiple controllers, I
+should be able to choose from them in defining my ``hardware-raid``
+configuration, or both.
+
+## Design Details
+
+- The CRD spec will have to be extended to add fields for ``physical_disks``
+  and ``controller`` under the ``hardware_raid`` section.
+- The provisioner will then be extended to process these fields.
+- There needs to be a map to transform user-interfacing names to the
+  vendor-specific names (like in case of ``firmware`` config).
```

Comment:
This is going to be a vendor-specific feature. How is it going to be handled?

Why use vendor-specific controller names but try to invent vendor-independent disk names?

Reply:
That's the unfortunate part where we don't have a purely vendor-agnostic interface available inside ironic that we can query through a singular format and extract information for each BMC type. What do you suggest otherwise, as a solution?